### PR TITLE
Validate if service has duplicate targetPort

### LINKED ISSUE #47222: Services allow duplicate targetPort
It is possible to express a Service where multiple front-side ports target the same back-side port.  This is not very useful and may cause trouble for implementations of Services.

```
apiVersion: v1
kind: Service
metadata:
  name: hostnames-2
spec:
  ports:
  - name: a
    port: 80
    targetPort: 9376
  - name: b
    port: 81
    targetPort: 9376
  selector:
    run: hostnames
````

### PR DESCRIPTION
**Pull request by xiangpengzhao**
**What this PR does / why we need it**:
Validate if a service has dup targetport

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #47222

**Special notes for your reviewer**:
/cc @thockin 
@kubernetes/sig-network-pr-reviews

**Release note**:

```release-note
NONE
```

### CODE REVIEW

File: pkg/api/validation/validation.go [hunk 1]
```diff
@@ -2865,6 +2865,21 @@ func ValidateService(service *api.Service) field.ErrorList {
 		nodePorts[key] = true
 	}
 
+	// Check for duplicate TargetPort
+	portsPath = specPath.Child("ports")
+	targetPorts := make(map[api.ServicePort]bool)
+	for i := range service.Spec.Ports {
```
Comment (odeke-em) [inline pkg/api/validation/validation.go]:
```go
for i, port := range service.Spec.Ports {
```

File: pkg/api/validation/validation.go [hunk 2]
```diff
@@ -2865,6 +2865,21 @@ func ValidateService(service *api.Service) field.ErrorList {
...
+		port := &service.Spec.Ports[i]
+		portPath := portsPath.Index(i)
+		var key api.ServicePort
```
Comment (odeke-em) [inline pkg/api/validation/validation.go]:
```go
key := api.ServicePort{TargetPort: port.TargetPort}
```

**Review by odeke-em — CHANGES_REQUESTED** [review]

File: pkg/api/validation/validation.go (hunk 1 above)
Comment (odeke-em) [inline pkg/api/validation/validation.go]: That way you don't have to do the `port := &service.Spec.Ports[i]`

**Comment by xiangpengzhao** [comment]
inline comments addressed. Thanks @odeke-em !

**Comment by feiskyer** [comment]
@k8s-bot ok to test

**Comment by kargakis** [comment]
@kubernetes/sig-network-api-reviews

File: pkg/api/validation/validation.go [hunk 3]
```diff
@@ -2865,6 +2865,19 @@ func ValidateService(service *api.Service) field.ErrorList {
 		nodePorts[key] = true
 	}
 
+	// Check for duplicate TargetPort
+	portsPath = specPath.Child("ports")
+	targetPorts := make(map[api.ServicePort]bool)
```
Comment (thockin) [inline pkg/api/validation/validation.go]: This should just be `map[intstr.IntOrString]bool` for obviousness reasons. or `struct{}` instead of bool.

**Review by thockin — COMMENTED** [review]
just a small point, then LGTM.

THANKS!!!

**Comment by thockin** [comment]
Tests fail because you're not special-casing the zero-value.  You should not de-dup number 0 or string ""
```
--- FAIL: TestServiceRegistryCreateMultiNodePortsService (0.00s)
	rest_test.go:230: Unexpected error: Service "foo1" is invalid: spec.ports[1].targetPort: Duplicate value: intstr.IntOrString{Type:0, IntVal:6503, StrVal:""}
FAIL
```

Assigning bowei to lgtm when done

/approve

**Comment by xiangpengzhao** [comment]
zero-value issue addressed. But I don't think it's the cause of the test failure. The failure was because I take "same targetport with different protocols" as dup and invalid. We should take this as valid, right?

File: pkg/api/validation/validation.go (hunk 3 above)
Comment (xiangpengzhao) [inline pkg/api/validation/validation.go]: We should also respect protocol, so I still use ServicePort here.

**Comment by xiangpengzhao** [comment]
@thockin @bowei PTAL.

**Pull request merged**